<a href="https://colab.research.google.com/github/BabakDavarmanesh/Learning_MachineLearning/blob/main/Machine_Learning_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color='darkblue'>Data Cleaning</font>

In most real-world scenarios, the data that data scientists work with is often raw, which means it may include duplicates, missing values, and irrelevant information. If this data is directly used for machine learning tasks, it can lead to poor accuracy or inaccurate predictions. Therefore, data cleaning—also referred to as data cleansing—is a crucial step that should be performed before any model development.

In this notebook, we will explore some common data cleaning techniques that data scientists typically use to preprocess and prepare their data for analysis.

## Objectives

*   Use Log function to transform the data
*   Handle the duplicates
*   Handle the missing values
*   Standardize and normalize the data
*   Handle the outliers


# **<font color='orange'>Import the required libraries</font>**


In [3]:
!pip install pandas numpy seaborn matplotlib scikit-learn scipy

In [4]:
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pylab as plt

from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler

from scipy.stats import norm
from scipy import stats

In this tutorial, we will be using the Ames_Housing_Data.tsv file, which you can download from [Kaggle](https://www.kaggle.com/datasets/hamzajabbarkhan/ames-housingtsv)

# **<font color='orange'>Exploring and making sense of our data</font>**

In [6]:
from google.colab import drive
drive.mount('/content/drive')

housing = pd.read_csv("/content/drive/My Drive/dataset/AmesHousing.tsv", sep='\t')  # Adjust path if needed
housing.head() # Default: Shows first 5 rows

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot Shape Land Contour  ... Pool Area Pool QC  Fence Misc Feature  \
0   NaN       IR1          Lvl  ...         0     NaN    NaN          NaN   
1   NaN       Reg          Lvl  ...         0     NaN  MnPrv          NaN   
2   NaN       IR1          Lvl  ...         0     NaN    NaN         Gar2   
3   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
4   NaN       IR1          Lvl  ...         0     NaN  MnPrv          NaN   

  Misc Val Mo Sold Yr Sold Sale Type  Sale Condition  SalePrice  
0        0       5    2010       WD           Normal     215000  
1        0       6    2010       WD           Normal     105000  
2    12500       6    2010       WD           Normal     172000  
3        0       4    2010       WD           Normal     244000  
4        0       3    2010       WD           Normal     189900  

[5 rows x 82 columns]

We can find more information about the features and types using the `info()`  method.

In [7]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order            2930 non-null   int64  
 1   PID              2930 non-null   int64  
 2   MS SubClass      2930 non-null   int64  
 3   MS Zoning        2930 non-null   object 
 4   Lot Frontage     2440 non-null   float64
 5   Lot Area         2930 non-null   int64  
 6   Street           2930 non-null   object 
 7   Alley            198 non-null    object 
 8   Lot Shape        2930 non-null   object 
 9   Land Contour     2930 non-null   object 
 10  Utilities        2930 non-null   object 
 11  Lot Config       2930 non-null   object 
 12  Land Slope       2930 non-null   object 
 13  Neighborhood     2930 non-null   object 
 14  Condition 1      2930 non-null   object 
 15  Condition 2      2930 non-null   object 
 16  Bldg Type        2930 non-null   object 
 17  House Style   

The dataset comprises 2,930 entries, indexed from 0 to 2,929, each characterized by 81 attributes.

The "Non-Null Count" column indicates the number of entries without missing values; if this count is 2,930 for a particular attribute, it signifies the absence of missing data for that attribute.

'SalePrice' serves as our target variable, while the remaining attributes act as predictor variables. The dataset contains a combination of numerical data types, with 28 columns of type int64 and 11 of type float64, as well as object data types.

In [8]:
housing["SalePrice"].describe()

count      2930.000000
mean     180796.060068
std       79886.692357
min       12789.000000
25%      129500.000000
50%      160000.000000
75%      213500.000000
max      755000.000000
Name: SalePrice, dtype: float64

​In this section, we'll explore the describe() function in pandas. This function provides a quick overview of the statistical properties of numerical data, which is essential for understanding the distribution and central tendencies of our target variable.​

Using `describe()` on 'SalePrice'

By applying `housing["SalePrice"].describe()`, we obtain the following statistics:​

- count: Total number of non-null entries.​

- mean: Average value of 'SalePrice'.​

- std: Standard deviation, indicating the variability around the mean.​

- min: Minimum value in the 'SalePrice' column.​

- 25%: First quartile (25th percentile), below which 25% of the data fall.​

- 50%: Median (50th percentile), the middle value separating the higher half from the lower half.​

- 75%: Third quartile (75th percentile), below which 75% of the data fall.​

- max: Maximum value in the 'SalePrice' column.​

These metrics collectively offer insights into the central tendency, spread, and range of housing sale prices in our dataset.